### PR TITLE
using ros-tool for battery indicator

### PR DESCRIPTION
**Pull request by jlurgo**
I'm using a node service (running within a local docker container) that simulates a robot battery slowly charging and discharging.

I added a battery charge indicator in the devices table.
I have yet to work on the abstractions I used to provide context to the indicator, there's a lot to extract into common components there.

WIP for #17 

Some screenshots:

While charging:
image
full:
image
medium:
image
low:
image
empty:
image

### CODE REVIEW

**Comment by chfritz** [comment]
Looks great!

> I have yet to work on the abstractions I used to provide context to the indicator, there's a lot to extract into common components there.

What do you have in mind?

File: src/client/components/battery-indicator.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    }
+  }, [isReady, subscribe])
+  
+  return <SimpleBatteryIndicator deviceData={deviceData} />
+}
+
+const BatteryIndicatorProvidedWithJwt = () => {
+  const jwt = useContext(JWTContext);
+  return(
+    <CapabilityContextProvider jwt={jwt}>
```
Comment (chfritz) [inline src/client/components/battery-indicator.tsx]: Please get
```
const host = import.meta.env.VITE_HOST; // Transitive deployment
const secure = !import.meta.env.VITE_INSECURE;
```
and then replace this line with `<CapabilityContextProvider jwt={jwt} host={host} ssl={secure}>`. This is needed for local development (when self-hosting Transitive).

**Comment by chfritz** [comment]
In the current approach we are starting a separate mqtt connection for each robot in the list. We can avoid that by using the fleet API provided by ROS Tool: https://transitiverobotics.com/caps/transitive-robotics/ros-tool/#fleet-api

File: src/client/components/battery-indicator.tsx (hunk 1 above)
Comment (jlurgo) [reply src/client/components/battery-indicator.tsx]: addressed

**Comment by jlurgo** [comment]
OK, I refactored the code so we use the fleet API
I also created some context providers to reuse ROS subscription/unsubscription code in context providers.

File: src/client/components/fleet-context.tsx [hunk 2]
```diff
@@ -13,46 +13,47 @@ const log = getLogger('FleetContext');
...
 const host = import.meta.env.VITE_HOST; // Transitive deployment
+const secure = !import.meta.env.VITE_INSECURE;
 const transitiveId = import.meta.env.VITE_TRANSITIVE_USER;
 const SSLs = import.meta.env.VITE_INSECURE ? '' : 's';
 const mqttUrl = `ws${SSLs}://mqtt.${host}`;
 
 export const FleetContext = createContext({});
 
-const ProviderWithJWT = ({ children }) => {
+const ProviderWithRos = ({ children }) => {
```
Comment (chfritz) [inline src/client/components/fleet-context.tsx]: Why do we have ProviderWithRos here and also ProviderWithROS in ros-provider?

**Comment by chfritz** [comment]
> I also created some context providers to reuse ROS subscription/unsubscription code in context providers.

I actually don't think that's a good idea. We want our users to use transAct to learn Transitive. We don't want them to learn transAct. What I mean by that is we don't want to introduce new wrappers inside transAct, e.g., by introducing new context providers around existing Transitive capabilities, in this case ROS Tool. If we do, then the user who looks at, say the Devices page code will first need to understand what the ROSContextProvider is and the API it exposes. And only once they do will they find the (documented) ROS Tool API underneath.

Instead we want to demonstrate how to use those APIs directly, and the simpler the example, the easier it will be for the user to understand. Another reason is that those APIs exposed by capabilities should be and are documented, e.g., [here](https://transitiverobotics.com/caps/transitive-robotics/ros-tool/#front-end-api) and any effort spent on documentation should be spent there, not on documenting another indirection we create inside transAct.

**Comment by jlurgo** [comment]
OK, I'll refactor it RN.
I was planning to propose these higher level components (or something more polished/reviewed) to be included as part of the Transitive API, along your **JWTContext** and **FleetContext**, but I'm realizing now that maybe I'm missing where the line between **Transitive** and **TransAct** is drawn. 
**JWTContext**, since it hits a **TransAct** endpoint (/api/getJWT) should be definitely on **TransAct** side. 
**FleetContext** depends on **JWTContext** and my multiple **ROSContext** depend on both, so they might be as well **TransAct** citizens.
I'll drop ROSContext

**Comment by chfritz** [comment]
Sounds good. And I definitely like the idea of adding more conveniences to the Transitive SDK when they generalize well to the varying setups that Transitive user might have.

**Review by chfritz — APPROVED** [review]

**Pull request merged**